kratos/containers/sparse_graph.h

```diff
@@ -0,0 +1,394 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Riccardo Rossi
+//
+
+#if !defined(KRATOS_SPARSE_GRAPH_H_INCLUDED )
+#define  KRATOS_SPARSE_GRAPH_H_INCLUDED
+
+
+// System includes
+#include <iostream>
+#include "includes/ublas_interface.h"
+#include "includes/serializer.h"
+
+// External includes
+#include <unordered_map>
+#include <unordered_set>
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+//class to construct and store a matrix graph. Can be used to construct efficiently a CSR matrix (or other sparse matrix types)
+
+/** This class is designed to store a matrix graph, aimed at the fast construction of other
+ * sparse matrix formats (particularly CSR)
+ * IMPORTANT NOTE: it is BY DESIGN NOT threadsafe! (a graph should be computed in each thread and then merged)
+*/
+
+class SparseGraph
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    typedef std::size_t IndexType;
+    typedef std::map<IndexType, std::unordered_set<IndexType> > GraphType; //using a map since we need it ordered
+
+    /// Pointer definition of SparseGraph
+    KRATOS_CLASS_POINTER_DEFINITION(SparseGraph);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    SparseGraph(){}
+
+    /// Destructor.
+    virtual ~SparseGraph(){}
+
+    /// Assignment operator. TODO: decide if we do want to allow it
+    SparseGraph& operator=(SparseGraph const& rOther)=delete;
+    // {
+    //     this->AddEntries(rOther.GetGraph());
+    //     return *this;
+    // }
+
+    /// Copy constructor. TODO: we need it otherwise the sendrecv does not work...
+    ///but i don't know why :-(
+    SparseGraph(const SparseGraph& rOther)
```

Comment:
Why not using the copy constructor of the mGraph?

Reply:
we can make it explicity, the point is that i don't understand why it is called in the sendrecv...